drivers: intel_adsp: IPC driver uses D3 power state

aborisovich opened the pull request:
This PR enables suspend/resume of the Intel ADSP IPC driver using Zephyr Power Manager Device API.
Options are enabled when CONFIG_PM_DEVICE=y.
New power control API performs hardware reinitialization after wake from D3 power state (PM_DEVICE_ACTION_SUSPEND):
* enabled interrupts
* resets register state to accept new messages
* allows to register callbacks to Zephyr application that may perform additional actions

Consists of 4 commits (2 main commits and two dependent):

1. Commit 2ea0d25892da45fdc316766efa52485b0ccdfade sets up Power Manager calls for IPC driver
2. Commit fa839a99d9113963a81a0a64d70887d4ceccb307 adds returned error codes to IPC functions, returning -EBUSY when Device is during power transition. 
3. Commit fc367d468ae86b8ba669b72e728645717fb8a795 allows system timer re-initialization - this commit is needed to resume system from low power state.
4. Commit 8dcac84c15ed3f7b1b8b553b1646e1204d4e53b2 allows to re-initialize memory window - this commit is also needed to complete IPC messages correctly after waking up from D3.

Blocked by bug:

- [x] https://github.com/zephyrproject-rtos/zephyr/issues/58502

Signed-off-by: Andrey Borisovich <andrey.borisovich@intel.com>

Commit:
intel_adsp: ipc: device support D3 power state change

When option CONFIG_PM_DEVICE is enabled, IPC driver is capable
of entering D3 power state described as PM_DEVICE_ACTION_SUSPEND
(and leaving that state - powering back to PM_DEVICE_ACTION_ACTIVE).
New power control callbacks 'ipc_power_control_api' are introduced for
use during power state transisions. They allow Zephyr application
specific code to be executed.

Signed-off-by: Andrey Borisovich <andrey.borisovich@intel.com>

Commit:
intel_adsp: mem_window: reinitialize after idle exit

Exiting idle state requires to reinitialize all memory window
instances to flush the cached memory.
Added function that calls initialization of devices during runtime.

Signed-off-by: Andrey Borisovich <andrey.borisovich@intel.com>

tmleman commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -315,6 +316,8 @@ __weak void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
 		z_xtensa_cache_flush_inv_all();
 		z_xt_ints_on(core_desc[cpu].intenable);
+
+		sys_clock_idle_exit();
```
I'm not sure but I think this should only be done for primary core and definitely only when IMR context saving is enabled.

tmleman replied:
What you are resolving issues without fixing them?

aborisovich replied:
Sorry forgot to push.

tmleman commented on soc/xtensa/intel_adsp/ace/power.c:
```diff
@@ -317,6 +318,7 @@ __weak void pm_state_exit_post_ops(enum pm_state state, uint8_t substate_id)
 		z_xtensa_cache_flush_inv_all();
 		z_xt_ints_on(core_desc[cpu].intenable);
 
+		mem_window_idle_exit();
```
Definitely only when IMR context saving is enabled and by the primary core only.

tmleman commented on soc/xtensa/intel_adsp/common/include/intel_adsp_ipc.h:
```diff
@@ -174,4 +178,38 @@ bool intel_adsp_ipc_send_message_sync(const struct device *dev,
+
+typedef int (*intel_adsp_ipc_suspend_handler_t)(const struct device *dev, void *arg);
+
+/**
+ * @brief Registers wake callback handler used to wake device from sleep.
+ *
+ * @param dev IPC device.
+ * @param fn Callback function.
+ * @param arg Value to pass as the "arg" parameter to the function.
+ */
+void intel_adsp_ipc_set_resume_handler(const struct device *dev,
+	intel_adsp_ipc_resume_handler_t fn, void *arg);
+
+/**
+ * @brief Registers suspend callback handler used to put device to sleep.
+ *
+ * @param dev IPC device.
+ * @param fn Callback function.
+ * @param arg Value to pass as the "arg" parameter to the function.
+ */
+void intel_adsp_ipc_set_suspend_handler(const struct device *dev,
+	intel_adsp_ipc_suspend_handler_t fn, void *arg);
```
This is not a blocker for me but I would avoid this two way dependency. The suspend callback is doing almost nothing and second one for resume is reinitializing edf. This operation can be done before FW Redy notification in notify_pm_state_exit.

aborisovich replied:
Ok, so there are few aspects here I would like to refer.

**Aspect 1 - proper code isolation and syntax**
We must have this two way dependency for now as the IPC driver code is split between Zephyr and SOF.  So, in any case we need to invoke IPC functions on the SOF side (like setting mask, initializing edf etc).
Now, you are suggesting that I should expose IPC driver private functions straight to the notify_pm_state_exit function.
I do not agree with this, since device initialization is a "private implementation" of the driver that should not be exposed outside. If we start to call private driver functions from everywhere bypassing interfaces and proper abstractions we would get a spaghetti code in no time.

**Aspect 2 - usefulness of the intel_adsp_ipc_suspend_handler**
* The `pm_policy_state_lock_is_active` lock used in the resume handler works even if power state is changed by force (with `pm_state_force()`)
* Given, that only IPC message is a source of signal for the whole system to enter D3 power state, we should validate consistency of IPC API. By his I mean, that when  `pm_state_force()` is called to enter D3, we must make sure that the `ipc->task_mask & IPC_TASK_POWERDOWN` is set in the SOF API part. Otherwise it is a fatal system condition.

ceolin replied:
This API is really confuse. Which device it is supposed to put to sleep ? The SoC ? If that is the case the whole patch is problematic and should not be done this way. The SoC should only be suspened from the idle thread.

Be careful with the nomemclature, from the Zephyr perspective device means peripherals.

aborisovich replied:
> Be careful with the nomemclature, from the Zephyr perspective device means peripherals.

From the Zephyr perspective, "Device" means a entity that implements [Device Driver Model](https://docs.zephyrproject.org/3.1.0/kernel/drivers/index.html), not peripherals. Maybe I forgot to write it with capital letter but in general my nomenclature is very precise. IPC is a Zephyr driver that implements Device Driver Model. This patch puts to sleep IPC Device.

> The SoC should only be suspened from the idle thread.

And it is. Unfortunately IPC driver API is split partially across SOF and Zephyr, that is why some things are very confusing.
This is done this way, because SOF maintains also IPC3 which does not use Zephyr IPC driver. SOF builds a façade on common IPC functions. This Zephyr IPC "API" is used from `sof\src\ipc\ipc-zephyr.c`.
To fully understand this PR, you should also have a look at the content of https://github.com/thesofproject/sof/pull/7325 .
Suspending logic goes like this:
1.  `sof\src\ipc\ipc-zephyr.c: ipc_platform_do_cmd()` function calls cpu_disable_core(PLATFORM_PRIMARY_CORE_ID) function. This sends instruction to Zephyr Power Manager, to enter D3 power state as soon as we enter idle (from idle thread). Feature now works for the primary core only, secondary cores D3 are under review too: https://github.com/zephyrproject-rtos/zephyr/pull/55182 .
2. Zephyr Power Manager during idle thread starts power state change (to PM_STATE_SOFT_OFF what we call D3).
3. Power state change invokes [PM notifiers](https://docs.zephyrproject.org/2.7.0/reference/power_management/index.html#structpm__notifier) callbacks when Power State is entered and exited.
Those are registered functions in `sof\src\platform\intel\ace\platform.c` (See https://github.com/thesofproject/sof/pull/7325/commits/defab6cf0bf1e70748a6b33f2bc4da26c9c4e7e5):
`static void notify_pm_state_entry(enum pm_state state)`
`static void notify_pm_state_exit(enum pm_state state)`
4. As we start to enter D3 state, `static void notify_pm_state_entry(enum pm_state state)` gets called. In this function body we prepare all Devices (and other entities) in the SoC to enter D3 state and do a context save (memory dump).
Function `pm_device_action_run(INTEL_ADSP_IPC_HOST_DEV, PM_DEVICE_ACTION_SUSPEND);` gets called that is a IPC Device PM API registered to [Zephyr Device Power Manager API](https://docs.zephyrproject.org/3.2.0/services/pm/device.html#device-model-with-power-management-support).
5. Calling `pm_device_action_run` on IPC Device, invokes the IPC Device power change handler:
`static int ipc_pm_action(const struct device *dev, enum pm_device_action action)`.
Inside this handler, we do following things:
    * mark IPC Device as busy - it now performs a power state change and should not be used. We set a proper status -  `pm_device_busy_set(dev);`.
    * take IPC Device PM lock - Device may not be asked to do power state change as it currently processes one:
`pm_device_state_lock(dev);`
    * At this point we invoke a callback to SOF IPC4 code as we need to validate the state of IPC subsystem.
    This is done using registered previously (with `intel_adsp_ipc_set_suspend_handler()`) function callback to: `sof\src\ipc\ipc-zephyr.c: ipc_device_suspend_handler()`
    * Inside this handler, we validate whether: IPC message queue is empty and `IPC_TASK_POWERDOWN` flag is set.
    If it is not, something is wrong with our internal messaging system and we terminate FW using assertion with proper error message indicating a reason.
    * If everything is ok, we release PM lock and set device state as not busy anymore. IPC Device power state had been changed to D3. The end.

Similar logic applies when we exit D3 state, just other notifier gets called, other callback to SOF IPC used and it does IPC Device reinitialization after waking up (enabling interrupts cleaning SOF task masks).

ceolin replied:
how are you calling `pm_device_action_run` ? I suspect you should be using device runtime. `pm_device_runtime_get` and `pm_device_runtime_put`   When device runtime is enabled the device is not suspended / resumed when the system and you do it manually, which is basically what you are doing. This will simplify this code a lot since you don't need to worry about ensuring that device is not suspended without you trigger it.

pjdobrowolski commented on soc/xtensa/intel_adsp/common/mem_window.c:
```diff
@@ -37,6 +37,14 @@ __imr int mem_win_init(const struct device *dev)
+	mem_win_init(DEVICE_DT_GET(MEM_WINDOW_NODE(0)));
+	mem_win_init(DEVICE_DT_GET(MEM_WINDOW_NODE(1)));
+	mem_win_init(DEVICE_DT_GET(MEM_WINDOW_NODE(2)));
+	mem_win_init(DEVICE_DT_GET(MEM_WINDOW_NODE(3)));
```
0-1-2-3 From where come that numbers?

aborisovich commented:
@nashif, @ceolin please review

ceolin commented on drivers/timer/intel_adsp_timer.c:
```diff
@@ -217,13 +226,23 @@ void smp_timer_init(void)
 
-SYS_INIT(sys_clock_driver_init, PRE_KERNEL_2,
-	 CONFIG_SYSTEM_CLOCK_INIT_PRIORITY);
+#ifdef CONFIG_PM_DEVICE
```
This does not seem to be right. This function is being called from system power management that only requires `CONFIG_PM` enabled.

aborisovich replied:
Good catch, thanks. I've tried to make it using [Device Power Management](https://docs.zephyrproject.org/3.2.0/services/pm/device.html#device-model-with-power-management-support) but dropped this idea. Forgot to adjust #ifdefs. I'll fix it in a moment,

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -205,12 +208,110 @@ static int dt_init(const struct device *dev)
+	/* All device pm is handled during resume and suspend */
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		pm_device_state_lock(dev);
```
Why are you locking the device state here and unlocking after calling `api->suspend_fn`. The state just changes if the system is suspended / resumed what only happens if idle thread runs.

aborisovich replied:
This is a validation mechanism provided by the Device Power API.

> The state just changes if the system is suspended / resumed what only happens if idle thread runs.

This is an assumption. Today, for MTL, this is done using only this use case ("what only happens if idle thread runs").
Tomorrow, there will be some additional use case, and system will behave in the uncontrolled way.
Device power state should be changed atomically.
That means, only if the transition of the power state had been successful, we change the Device power state.
That is why, we take a `pm_device_state_lock` and release the lock just before successful completion of the transition logic.
If the Device fails to change power state (and assertion is called), its power state should remain unchanged (this sentence looks weird but well...).
Quoting docs of the [pm_device_state_lock](https://docs.zephyrproject.org/apidoc/latest/group__subsys__pm__device.html#gaab27e932950e1063b2f1f4c4e19dbf01):

> This function locks the current device power state. Once locked the device power state will not be changed by system power management or device runtime power management until unlocked.

ceolin replied:
If you really don't want the system doing power management in devices and have to do it completely manually, please enable `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE` and use device runtime. That is cleanest way to do it.

ceolin replied:
Can you explaing when this function is preempted by the idle thread ? 

This whole logic is not making sense for me, if you manually controlling power management for this device, you should *not* do it using `pm_device_action_run` but through `pm_device_runtime_get/put`. The latter ensure that the system won't interrupt an operation and do the bookkepping of who is using the driver ... Manually invoke `pm_device_action_run` and use system power management is not right. This callback should just care about do the action it was called and not care about locking, state and other things.

aborisovich replied:
> If you really don't want the system doing power management in devices and have to do it completely manually, please enable `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE` and use device runtime. That is cleanest way to do it.

No it is not. I've been there and tried to implement this initially. Also @tmleman initially proposed usage of `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE`.
* Device Runtime Power API is used for devices that are put to sleep when not in use. This is not suitable for IPC Device since it should be active at all times when SoC is awake.
* We decided to call manually `pm_device_action_run` functions in the notifiers, because mode simple [Central Method](https://docs.zephyrproject.org/2.6.0/reference/power_management/index.html#central-method) does not meet our needs. In this method, calling function `pm_system_suspend()` would result in the putting to sleep add Devices in the system to D3, but:
    1. We can not determine order of this action
    2. We can not provide any error handling inside this operation (assertions, error codes, etc)
    3. `pm_system_suspend` just ignores devices that are busy:
    > If a decision to enter deep sleep is made, the implementation would enter it only after checking if the devices are not in the middle of a hardware transaction that cannot be interrupted.
    
    This is not a right solution in our case - we receive IPC instructing us to enter D3 and we should just enter it.
    We can optionally take some time waiting for some devices to finish hardware operations and exit busy state,
    but calling `pm_system_suspend()` in a loop until it succeeds seems a bad design to me.
    This Central Method is good for small embedded systems not for our firmware (or at least its current form).

aborisovich replied:
> Can you explaing when this function is preempted by the idle thread ?
> 
> This whole logic is not making sense for me, if you manually controlling power management for this device, you should _not_ do it using `pm_device_action_run` but through `pm_device_runtime_get/put`. The latter ensure that the system won't interrupt an operation and do the bookkepping of who is using the driver ... Manually invoke `pm_device_action_run` and use system power management is not right. This callback should just care about do the action it was called and not care about locking, state and other things.

This is simply not a true statement. Function [pm_device_state_set](https://docs.zephyrproject.org/2.6.0/reference/power_management/index.html#device-set-power-state) and [pm_device_state_get](https://docs.zephyrproject.org/2.6.0/reference/power_management/index.html#device-get-power-state) are part of the Zephyr Power Manager **public interface** to be used exactly in the way I do and exactly in the scenario I have (when using [Central Method](https://docs.zephyrproject.org/2.6.0/reference/power_management/index.html#central-method) is not enough).
I've explained why we should not use [Device Runtime Power Management](https://docs.zephyrproject.org/2.6.0/reference/power_management/index.html#device-runtime-power-management) for IPC Device in previous comments.

ceolin replied:
@aborisovich

>Device Runtime Power API is used for devices that are put to sleep when not in use. This is not suitable for IPC Device since it should be active at all times when SoC is awake.

Yes,because some drivers do pm_device_runtime_get/put(self) when do an operation. That is not the case of this device. In this device if you use `CONFIG_PM_DEVICE_RUNTIME` the device will *only* be suspended when you call `pm_device_runtime_put` that is done only when you need. That is exactly alternative for the "Central Method".

>We decided to call manually pm_device_action_run functions in the notifiers, because more simple Central Method does not meet our needs. In this method, calling function pm_system_suspend() would result in the putting to sleep add Devices in the system to D3, but:

That is the whole point of device runtime power management. If pm_device_runtime_get/put is reference counted. If you do a `pm_device_runtime_get` the device will be active until you do `pm_device_runtime_put` and there is no other place holding a reference. e.g: You call `pm_device_runtime_get`, another place calls `pm_device_runtime_get` do something and call `pm_device_runtime_put`, the device will sill be active because the referece count is still 1.

> This is not a right solution in our case - we receive IPC instructing us to enter D3 and we should just enter it. We can optionally take some time waiting for some devices to finish hardware operations and exit busy state, but calling pm_system_suspend() in a loop until it succeeds seems a bad design to me. This Central Method is good for small embedded systems not for our firmware (or at least its current form).

Yes, so why you are relying on this method ? You should be using device runtime pm :)

> This is simply not a true statement. Function pm_device_state_set and pm_device_state_get are part of the Zephyr Power Manager public interface to be used exactly in the way I do and exactly in the scenario I have (when using Central Method is not enough). I've explained why we should not use Device Runtime Power Management for IPC Device in previous comments.

Yes, that is true and we probably need to document it better to not recommend using it directly as you are doing. And the problem is that from the OS perspective if applications use it directly is hard to keep the system sane. Your use case (if I am not understanding wrongly) is the exactly use case for device runtime power management.

ceolin replied:
@aborisovich

>Device Runtime Power API is used for devices that are put to sleep when not in use. This is not suitable for IPC Device since it should be active at all times when SoC is awake.

Yes,because some drivers do pm_device_runtime_get/put(self) when do an operation. That is not the case of this device. In this device if you use `CONFIG_PM_DEVICE_RUNTIME` the device will *only* be suspended when you call `pm_device_runtime_put` that is done only when you need. That is exactly alternative for the "Central Method".

>We decided to call manually pm_device_action_run functions in the notifiers, because more simple Central Method does not meet our needs. In this method, calling function pm_system_suspend() would result in the putting to sleep add Devices in the system to D3, but:

That is the whole point of device runtime power management. If pm_device_runtime_get/put is reference counted. If you do a `pm_device_runtime_get` the device will be active until you do `pm_device_runtime_put` and there is no other place holding a reference. e.g: You call `pm_device_runtime_get`, another place calls `pm_device_runtime_get` do something and call `pm_device_runtime_put`, the device will sill be active because the referece count is still 1.

> This is not a right solution in our case - we receive IPC instructing us to enter D3 and we should just enter it. We can optionally take some time waiting for some devices to finish hardware operations and exit busy state, but calling pm_system_suspend() in a loop until it succeeds seems a bad design to me. This Central Method is good for small embedded systems not for our firmware (or at least its current form).

Yes, so why you are relying on this method ? You should be using device runtime pm :)

> This is simply not a true statement. Function pm_device_state_set and pm_device_state_get are part of the Zephyr Power Manager public interface to be used exactly in the way I do and exactly in the scenario I have (when using Central Method is not enough). I've explained why we should not use Device Runtime Power Management for IPC Device in previous comments.

Yes, that is true and we probably need to document it better to not recommend using it directly as you are doing. And the problem is that from the OS perspective if applications use it directly is hard to keep the system sane. Your use case (if I am not understanding wrongly) is the exactly use case for device runtime power management.

ceolin replied:
@aborisovich

>Device Runtime Power API is used for devices that are put to sleep when not in use. This is not suitable for IPC Device since it should be active at all times when SoC is awake.

Yes,because some drivers do pm_device_runtime_get/put(self) when do an operation. That is not the case of this device. In this device if you use `CONFIG_PM_DEVICE_RUNTIME` the device will *only* be suspended when you call `pm_device_runtime_put` that is done only when you need. That is exactly alternative for the "Central Method".

>We decided to call manually pm_device_action_run functions in the notifiers, because more simple Central Method does not meet our needs. In this method, calling function pm_system_suspend() would result in the putting to sleep add Devices in the system to D3, but:

That is the whole point of device runtime power management. If pm_device_runtime_get/put is reference counted. If you do a `pm_device_runtime_get` the device will be active until you do `pm_device_runtime_put` and there is no other place holding a reference. e.g: You call `pm_device_runtime_get`, another place calls `pm_device_runtime_get` do something and call `pm_device_runtime_put`, the device will sill be active because the referece count is still 1.

> This is not a right solution in our case - we receive IPC instructing us to enter D3 and we should just enter it. We can optionally take some time waiting for some devices to finish hardware operations and exit busy state, but calling pm_system_suspend() in a loop until it succeeds seems a bad design to me. This Central Method is good for small embedded systems not for our firmware (or at least its current form).

Yes, so why you are relying on this method ? You should be using device runtime pm :)

> This is simply not a true statement. Function pm_device_state_set and pm_device_state_get are part of the Zephyr Power Manager public interface to be used exactly in the way I do and exactly in the scenario I have (when using Central Method is not enough). I've explained why we should not use Device Runtime Power Management for IPC Device in previous comments.

Yes, that is true and we probably need to document it better to not recommend using it directly as you are doing. And the problem is that from the OS perspective if applications use it directly is hard to keep the system sane. Your use case (if I am not understanding wrongly) is the exactly use case for device runtime power management.

ceolin replied:
> Tomorrow, there will be some additional use case, and system will behave in the uncontrolled way.
Device power state should be changed atomically.

You have to be specific about the use case that you want to cover. Also, what you are doing is not atomic you can be preempted in the middle of this operation and in that case you may be in an inconsistent state if you try to use this device.

THAT assuming that you are suspending / resuming this device out of the idle loop.

My point is, if you are trying to cover that other places in the code will suspend or resume this device, this operation is not being atomic or if you know that suspend/resume will happen only from the idle thread (or using the proper device runtime API) this lock/unlock mechanism is not needed ...

The purpose of this API is to prevent other places in the code of changing the device state and *NOT* to guarantee state transition atomicity.

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -205,12 +208,110 @@ static int dt_init(const struct device *dev)
+	switch (action) {
+	case PM_DEVICE_ACTION_SUSPEND:
+		pm_device_state_lock(dev);
+		pm_device_busy_set(dev);
```
Why locking the current state and set it as busy ?

aborisovich replied:
As of `pm_device_state_lock` usage, please see my https://github.com/zephyrproject-rtos/zephyr/pull/56168#discussion_r1171105609 answer.
However, good catch on the `pm_device_busy_set()`. Looks like I misused this mechanism as it has other purpose that I though. Turns out it is used to **prevent** power state changes when eg. hardware operation is performed.
From the [Device Power Manager - Busy indication](https://docs.zephyrproject.org/3.2.0/services/pm/device.html#busy-status-indication):

> (...) Suspending a device which is in the middle of a hardware transaction, such as writing to a flash memory, may lead to undefined behavior or inconsistent states. This API guards such transactions by indicating to the kernel that the device is in the middle of an operation and should not be suspended.

I'll remove this busy indicator (its purpose is different). Thanks.

aborisovich replied:
Fixed, implementation changed. Moved Device busy indicators to IPC transactions where registers are written.

ceolin commented:
It is not clear for me what you are trying to do, if you want to only suspend a peripheral when you get an IPC message or if you want to suspend the whole CPU when you get this message

ceolin commented on drivers/timer/intel_adsp_timer.c:
```diff
@@ -217,13 +226,23 @@ void smp_timer_init(void)
+#ifdef CONFIG_PM_DEVICE
+void sys_clock_idle_exit(void)
+{
+	pm_device_state_lock(DEVICE_DT_INST_GET(0));
```
These `state_lock|unlock` functions don't seems to be necessary

aborisovich replied:
Yes, we do not use here Device Power Manager model, so there is no such thing as device power state.
I'll remove them in a moment.

ceolin reviewed (changes requested):
Many things to clarify, -1 to prevent it to be merged.

aborisovich commented:
Revision update - addressed remarks from @ceolin code review:
* removed `pm_device_state_lock` from `drivers/timer/intel_adsp_timer.c`
* changed `#ifdef CONFIG_PM_DEVICE` to `#ifdef CONFIG_PM` in `drivers/timer/intel_adsp_timer.c`
* removed `pm_device_busy_set/pm_device_busy_clear` from `static int ipc_pm_action(const struct device *dev, enum pm_device_action action)` function where it was not needed (incorrect usage)
* changed how `pm_device_state_lock` works so it makes more sense - if any function in the body of `ipc_pm_action()` (single power state transition) returns error code, it is returned by the `pm_device_action_run` and the lock is not removed. Lock is removed only when power transition completes successfuly.
* added documentation about the PM lock to the `pm_device_action_run` function

New changes:
* added functions `pm_device_busy_set()` and `pm_device_busy_clear()` for IPC Device in functions where registers are set so the function is not interrupted in the middle by the Zephyr Power Manager power change. This is not directly required by SOF MTL D3 implementation since it assumes, that power state may be changed only from the idle thread. However due to generic nature of the implementation and API this behavior is beneficial.

aborisovich commented:
Revision update:
* codestyle

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -102,18 +109,29 @@ int intel_adsp_ipc_init(const struct device *dev)
+
+#ifdef CONFIG_PM_DEVICE
+	pm_device_busy_clear(dev);
+#endif /* CONFIG_PM_DEVICE */
```
Why this busy_set/clear here ? I don't get the point, this function is not doing anything that would block and possibly let the idle thread runs ... Am I missing something ?

aborisovich replied:
Those `pm_device_busy_clear`/`pm_device_busy_set` indicate whether Device is busy or not.
Sorry for copy paste of previous answer, I'll just explain what I mean:
> From the [Device Power Manager - Busy indication](https://docs.zephyrproject.org/3.2.0/services/pm/device.html#busy-status-indication):
> 
> > (...) Suspending a device which is in the middle of a hardware transaction, such as writing to a flash memory, may lead to undefined behavior or inconsistent states. This API guards such transactions by indicating to the kernel that the device is in the middle of an operation and should not be suspended.

When we start writing IPC registers, this is a part of "hardware transaction" between DSP and Host.
We do this during sending IPC message and acknowledging it. From MTL SoC design, we assume that there may be no running transactions between Host and DSP, because we enter D3 state during idle.
However, existing IPC messaging scheduling works using [k_work_schedule](https://docs.zephyrproject.org/3.1.0/kernel/services/threads/workqueue.html#c.k_work_schedule). We put IPC tasks in SOF on the queue and those tasks are scheduled for execution in some later time.
Theoretically, it is possible to enter idle thread and cause power transition while there are still IPC messages in the queue (that were scheduled for later execution). So in practice it may happen, that we are performing a power transition while IPC message is being send and registers are written. Those busy indicators are safeguards for this case.
If the very unlikely scenario mentioned above would happen, thanks to those safeguards a call to  `pm_device_action_run(INTEL_ADSP_IPC_HOST_DEV, PM_DEVICE_ACTION_SUSPEND);` would result in a Power Manager ignoring this device and not performing transition.
I've just noticed that this is valid only if devices are put to sleep using `pm_suspend_devices` function.
Direct call to the `pm_device_action_run()` ignores Device busy state anyway. I'll add some checks on the SOF side to platform pm_notifiers.

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -102,18 +109,29 @@ int intel_adsp_ipc_init(const struct device *dev)
+
+#ifdef CONFIG_PM_DEVICE
+	pm_device_busy_clear(dev);
+#endif /* CONFIG_PM_DEVICE */
```
ditto

aborisovich commented:
Revision update - answered @ceolin review questions:
* added new `wait_ipc_device_not_busy` function that waits for 5 seconds in total for IPC device to become non-busy.
* updated `ipc_pm_action()` function with a call to `wait_ipc_device_not_busy()` as a safeguard to prevent power gating when there are still IPC messages processing.

When `ipc_pm_action()` is invoked by the Zephyr Power Manager function `pm_system_suspend()`, Devices with state busy are ignored and their power state is not changed. However, when power state change of the Device is invoked directly using
`pm_device_action_run()`, Device busy status is ignored. Added additional safeguards in the `ipc_pm_action()`.

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -205,12 +218,157 @@ static int dt_init(const struct device *dev)
+
+	/* wait up to 5 seconds for IPC device to become non-busy */
+	if(!wait_ipc_device_not_busy(1000000, 5))
+		return -EBUSY;
```
I see this logic here, do you need `intel_adsp_ipc_resume_handler_t`  and `intel_adsp_ipc_suspend_handler_t` ? My understand that that API is because the IPC message can vary depending on target, but isn't this block the same ?

Can't we make simple it and having the msg logic in this driver instead of that callback ? That is a pattern that is not common in our drivers, maybe the right thing to do is having a kernel event for when the device change the status and the application has a listener for it where it does whathever it wants instead of doing it inside the pm action.

aborisovich replied:
@ceolin a right thing to do is to move IPC driver to Zephyr completely. However, this is not possible now until IPC3 exists (that does not use this Zephyr driver). Callbacks you mentioned are defined just because we can not #include sof headers in the Zephyr code and some key IPC elements are in the SOF.
So we do this "the hard way":
* define callback register function in Zephyr
* call callback register function in SOF and assign some SOF IPC function to it.
* Zephyr driver in some point calls callback.
* thanks to this, steering returns to SOF (some IPC function)
* then we do what we need to do (and we reference elements that are defined on SOF side eg. ipc message queue, ipc flags, ipc tasks)

Take a look how other callbacks are registered in this API (eg. isr). Adding two more is nothing new really.

ceolin replied:
That is what I thought, understand the point and I am fine with that.

ceolin commented on drivers/timer/intel_adsp_timer.c:
```diff
@@ -214,16 +220,24 @@ void smp_timer_init(void)
+}
+#endif /* CONFIG_PM */
+
+DEVICE_DT_INST_DEFINE(0, sys_clock_driver_init, NULL, NULL, NULL,
```
Why have you change it ? That makes this driver different from others sys_clock*. That is a pretty weird driver that and uses SYS_INIT and has some peculiarities as far as I remember. This change does not seem necessary since you don't need the device pointer for anything. Keep it simple and don't change it please.

aborisovich replied:
I do not change things that are not mandatory.
We need Device instance pointer to pass it to the clock init function. Previously this function was statically initialized with macros, now I need to call it run-time with proper argument.

ceolin replied:
```
static int sys_clock_driver_init(const struct device *dev)
{
	uint64_t curr = count();
	if (!sys_clock_driver_reinit)
		IRQ_CONNECT(TIMER_IRQ, 0, compare_isr, 0, 0);
	set_compare(curr + CYC_PER_TICK);
	last_count = curr;
	irq_init();
	return 0;
}
```

Where `dev` is used there ? You can simply call `sys_clock_driver_init(NULL)`.

aborisovich replied:
Yes... Indeed I can. Call will look a bit weird but good catch. I will fix it tomorrow, thanks.

ceolin replied:
It is better than this be the only one doing it IMHO :)

ceolin commented on drivers/timer/intel_adsp_timer.c:
```diff
@@ -214,16 +220,24 @@ void smp_timer_init(void)
 
-SYS_INIT(sys_clock_driver_init, PRE_KERNEL_2,
-	 CONFIG_SYSTEM_CLOCK_INIT_PRIORITY);
+#ifdef CONFIG_PM
```
remove this ifdef, that is regular API for sys_clock and you are not using anything that requires `CONFIG_PM`. If this function is not called the linker you remove it. We rely on it for this sort of thing.

aborisovich replied:
Restored to previous SYS_INIT macro

nashif commented on drivers/timer/intel_adsp_timer.c:
```diff
@@ -214,16 +220,24 @@ void smp_timer_init(void)
 	uint64_t curr = count();
-
-	IRQ_CONNECT(TIMER_IRQ, 0, compare_isr, 0, 0);
+	if (!sys_clock_driver_reinit)
```
braces

aborisovich commented:
Marked as DNM. We discussed with @tmleman that it would be better to move `pm_device_action_run` for IPC device to _soc\xtensa\intel_adsp\ace\power.c_ `pm_state_set`/`pm_state_exit_post_ops` where actual power transition takes place.

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -205,12 +218,157 @@ static int dt_init(const struct device *dev)
+	.suspend_fn_args = NULL
+};
+
+PM_DEVICE_DT_DEFINE(INTEL_ADSP_IPC_HOST_DTNODE, ipc_pm_action);
```
This could be out of this `#ifdef`
` @note This macro is a no-op if @kconfig{CONFIG_PM_DEVICE} is not enabled.` same for `PM_DEVICE_DT_GET`.
That could simplify it. I know that you have an additional parameter that is passed only this option is enabled,
but you still could have something like:
```
DEVICE_DT_DEFINE(INTEL_ADSP_IPC_HOST_DTNODE, dt_init, PM_DEVICE_DT_GET(INTEL_ADSP_IPC_HOST_DTNODE),
	&ipc_host_data, &ipc_host_config, PRE_KERNEL_2, 0, COND_CODE_1(CONFIG_PM_DEVICE, (&ipc_power_control_apii), (NULL)));
```

ceolin commented:
@aborisovich

>Device Runtime Power API is used for devices that are put to sleep when not in use. This is not suitable for IPC Device since it should be active at all times when SoC is awake.

Yes,because some drivers do pm_device_runtime_get/put(self) when do an operation. That is not the case of this device. In this device if you use `CONFIG_PM_DEVICE_RUNTIME` the device will *only* be suspended when you call `pm_device_runtime_put` that is done only when you need. That is exactly alternative for the "Central Method".

>We decided to call manually pm_device_action_run functions in the notifiers, because more simple Central Method does not meet our needs. In this method, calling function pm_system_suspend() would result in the putting to sleep add Devices in the system to D3, but:

That is the whole point of device runtime power management. If pm_device_runtime_get/put is reference counted. If you do a `pm_device_runtime_get` the device will be active until you do `pm_device_runtime_put` and there is no other place holding a reference. e.g: You call `pm_device_runtime_get`, another place calls `pm_device_runtime_get` do something and call `pm_device_runtime_put`, the device will sill be active because the referece count is still 1.

> This is not a right solution in our case - we receive IPC instructing us to enter D3 and we should just enter it. We can optionally take some time waiting for some devices to finish hardware operations and exit busy state, but calling pm_system_suspend() in a loop until it succeeds seems a bad design to me. This Central Method is good for small embedded systems not for our firmware (or at least its current form).

Yes, so why you are relying on this method ? You should be using device runtime pm :)

> This is simply not a true statement. Function pm_device_state_set and pm_device_state_get are part of the Zephyr Power Manager public interface to be used exactly in the way I do and exactly in the scenario I have (when using Central Method is not enough). I've explained why we should not use Device Runtime Power Management for IPC Device in previous comments.

Yes, that is true and we probably need to document it better to not recommend using it directly as you are doing. And the problem is that from the OS perspective if applications use it directly is hard to keep the system sane. Your use case (if I am not understanding wrongly) is the exactly use case for device runtime power management.

tmleman commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -205,12 +218,157 @@ static int dt_init(const struct device *dev)
+{
+	if (pm_device_is_busy(INTEL_ADSP_IPC_HOST_DEV)) {
+		while (retries > 0) {
+			k_busy_wait(usec_to_wait);
```
I am not sure if this will work properly. Since we're waiting in idle for another thread to be scheduled, is using _busy_wait_ a good idea here? Maybe _yield_ or _sleep_ would be better?

aborisovich replied:
Yes, it is bad idea now since we established to use native Device Power Manager and expect devices to be suspended automatically by kernel (using `pm_system_suspend()`).
`pm_system_suspend()` existing implementation will just ignore device that are busy. Moreover, idle thead should not be blocking. I've consulted @ceolin on this, we should just make sure that power transition of device is not started when device is busy.

andyross commented on drivers/timer/intel_adsp_timer.c:
```diff
@@ -217,13 +223,22 @@ void smp_timer_init(void)
-	IRQ_CONNECT(TIMER_IRQ, 0, compare_isr, 0, 0);
+	if (!sys_clock_driver_reinit) {
+		IRQ_CONNECT(TIMER_IRQ, 0, compare_isr, 0, 0);
+	}
```
This seems odd, what is it trying to fix?  IRQ_CONNECT() isn't code[1], it's a build-time construct that associates an ISR with an interrupt by emitting a record to a special linker section.

This is a noop, basically.  Are you maybe wanting to control irq_en/disable() instead?

[1] On most architectures, including xtensa.  There are a few that require something happen at runtime, but I'd have to look up the details.

aborisovich replied:
Sorry, the documentation of `#define IRQ_CONNECT`:

> Although this routine is invoked at run-time, all of its arguments must be
>  * computable by the compiler at build time.
>  * (...)

Made me think, that this function is called runtime.

aborisovich replied:
Yes, this is build-time construct as you said.
Left just the `sys_clock_idle_exit()` function that simply calls `sys_clock_driver_init()`.

fkwasowi commented:
The problem also applies to runtime_idle state, not only D3. Request for porting https://github.com/intel-innersource/os.rtos.zephyr.zephyr/pull/679

aborisovich commented:
Push update:

- removed call to `wait_ipc_device_not_busy()` from the `ipc_pm_action()` function - idle thread should not be blocking
- function `wait_ipc_device_not_busy()` will be used on the SOF side to make sure we do not start power transition during IPC message pending
- cherry-picked @fkwasowi commit
- added for testing commit from https://github.com/zephyrproject-rtos/zephyr/pull/58245 - this allows for IPC device to be suspended by the kernel automatically (without explicit call to pm_device_action_run(INTEL_ADSP_IPC_HOST_DEV, PM_DEVICE_ACTION_SUSPEND) on SOF side).
- alterned device pm busy flag - now it is used more like it should be. There are to "ranges" where IPC device is busy:
    1. during initialization
    2. from sending message to host to acknowledgement of the message by host using ISR. It could theoretically replace a `devdata->tx_ack_pending` flag, but we are going to support IPC device without PM_DEVICE enabled.

aborisovich commented:
Closed by misclick.

Commit:
inte_adsp: ipc: prevent ipc message send during Device power transition

When CONFIG_PM_DEVICE is enabled IPC Device may be during power transition
during a call to intel_adsp_ipc_send_message function.
Changed signatures of intel_adsp_ipc_send_message and its sync version
to return int and negative error codes on error.
On attempt to send IPC message during Device power transition
-ESHUTDOWN error code is returned, on busy state -EBUSY.
Updated all function references.

Signed-off-by: Andrey Borisovich <andrey.borisovich@intel.com>

lyakh commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -207,12 +216,149 @@ static int dt_init(const struct device *dev)
+ * @return int Returns 0 on success or optionaly error code from the
+ * registered ipc_power_control_api callbacks. Returning error code from functions
+ * invoked during power stansition, does not release the pm_device_state_lock.
+ * It must be released manually by the caller during error handling process.
```
this is strange, any particular reason for this?

aborisovich replied:
No, comment I forgot to remove from older revisions of this PR. I'll clean it up.

Commit:
intel_adsp: timer: implemented sys_clock_idle_exit function

Generic header for system clock allows to define a sys_clock_idle_exit
function for the clock implementation.
Implemented the function in the intel_adsp_timer to reinitialize
device driver after the idle exit state.

Signed-off-by: Andrey Borisovich <andrey.borisovich@intel.com>

marc-hb reviewed (commented):
https://github.com/zephyrproject-rtos/zephyr/actions/runs/5175134859/jobs/9322425984?pr=56168 failed to build with:

```
zephyr/zephyr/soc/xtensa/intel_adsp/common/ipc.c:262:13:
  error: 'wait_ipc_device_not_busy' defined but not used [-Werror=unused-function]
  262 | static bool wait_ipc_device_not_busy(uint32_t usec_to_wait, uint32_t retries)
      |             ^~~~~~~~~~~~~~~~~~~~~~~~
cc1: all warnings being treated as errors
```

aborisovich commented:
> https://github.com/zephyrproject-rtos/zephyr/actions/runs/5175134859/jobs/9322425984?pr=56168 failed to build with:
> 
> ```
> zephyr/zephyr/soc/xtensa/intel_adsp/common/ipc.c:262:13:
>   error: 'wait_ipc_device_not_busy' defined but not used [-Werror=unused-function]
>   262 | static bool wait_ipc_device_not_busy(uint32_t usec_to_wait, uint32_t retries)
>       |             ^~~~~~~~~~~~~~~~~~~~~~~~
> cc1: all warnings being treated as errors
> ```

Fixed, all CI green.

aborisovich commented:
Last iteration update:
* rebased to newest Zephyr main what fixed the #58502 bug and allowed to use Zephyr System Power Manager to suspend IPC device.
* removed `wait_ipc_device_not_busy()` function as @tmleman suggested. Moreover it is not needed anymore since we use default Zephyr kernel PM behavior - function `pm_system_suspend` that is responsible for suspending devices.
* removed conditional `IRQ_CONNECT` from `intel_adsp_timer.c::sys_clock_idle_exit()` function as it is build-time construct (addressed @andyross remark).
* added new commit fa839a99d9113963a81a0a64d70887d4ceccb307 that changes signature of the `intel_adsp_ipc_send_message()` function (previously was returning `bool`, now `int` with return code). This was done to return different values when IPC device is busy with pending message handling or is during power transition.

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -207,12 +219,123 @@ static int dt_init(const struct device *dev)
+		pm_device_state_lock(dev);
+		irq_enable(DT_IRQN(INTEL_ADSP_IPC_HOST_DTNODE));
+		if (!irq_is_enabled(DT_IRQN(INTEL_ADSP_IPC_HOST_DTNODE))) {
+			ret = -ECANCELED;
```
`ECANCELLED` is pretty weird for this error ... which operation was cancelled ? You are enabling an irq line.

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -207,12 +219,123 @@ static int dt_init(const struct device *dev)
+		break;
+	case PM_DEVICE_ACTION_RESUME:
+		pm_device_state_lock(dev);
+		irq_enable(DT_IRQN(INTEL_ADSP_IPC_HOST_DTNODE));
```
where this irq line was disabled ?

aborisovich replied:
I thought System PM will do this for the Device. I'll add it.

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -207,12 +219,123 @@ static int dt_init(const struct device *dev)
+		pm_device_state_lock(dev);
+		if (api->suspend_fn) {
+			ret = api->suspend_fn(dev, api->suspend_fn_args);
+			if (ret)
```
missing braces

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -207,12 +219,123 @@ static int dt_init(const struct device *dev)
+		if (api->suspend_fn) {
+			ret = api->suspend_fn(dev, api->suspend_fn_args);
+			if (ret)
+				break;
```
What happens here ? state will continue locked ? The system won't be able to recover from it since pm_device_action_run checks if the device's state is locked before trigger the action callback.

aborisovich commented:
Push update:
* addressed @ceolin review. Is returning `-EINTR` code better when `!irq_is_enabled()`? Could not find more suitable EC.

aborisovich commented:
> @aborisovich did you see these comments [#56168 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/56168#discussion_r1218892512)

Yes. Ok, from my perspective, Zephyr System Power Management API exposes `pm_device_action_run()` as part of public API.
There are some ideas to share this IPC driver to other architectures not only Intel ADSP.
`pm_device_action_run()` function allows to change power state of the driver without using idle thread.
That is why those locks are still here. I'm trying to write a more generic code that would be usable also when somebody calls
`pm_device_action_run(INTEL_ADSP_IPC_HOST_DEV)`. Then the device would not change power state if it is locked.

ceolin commented:
> > @aborisovich did you see these comments [#56168 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/56168#discussion_r1218892512)
> 
> Yes. Ok, from my perspective, Zephyr System Power Management API exposes `pm_device_action_run()` as part of public API. There are some ideas to share this IPC driver to other architectures not only Intel ADSP. `pm_device_action_run()` function allows to change power state of the driver without using idle thread. That is why those locks are still here. I'm trying to write a more generic code that would be usable also when somebody calls `pm_device_action_run(INTEL_ADSP_IPC_HOST_DEV)`. Then the device would not change power state if it is locked (because it currently performs a transaction).
> 
> As for the `device_busy` lock, consider following scenario:
> 
> 1. Interrupt from Host with IPC message send to DSP.
> 2. We schedule IPC answer to Host.
> 3. Executing IPC send, message is send but receival is not confirmed by the Host.
> 4. We go back to idle thread waiting for another interrupt from Host that will confirm IPC message receival.
> 5. ... and we go to sleep since we processed D3 command and are in the idle tread.
> 
> This lock prevents this from happening. I understand that "this should not happen since Host responds fast". But what if it don't?

You are getting it wrong. I am not assuming anything about timing .... I am contesting the efficacy of this lock. You are using a lock to ensure that the device state changing will happen atomically and that is not the case. Following your steps:
> 1.  Interrupt from Host with IPC message send to DSP
> 2.  We schedule IPC answer to Host.

 ---> IPC device is "active" at this point right ? You don't schedule this ipc from inside pm_device_action_run(), correct ?
       So, what you should do here, is from the function you are scheduling the ipc you set the device busy

> 3. Executing IPC send, message is send but receival is not confirmed by the Host.
> 4.  We go back to idle thread waiting for another interrupt from Host that will confirm IPC message receival.
> 5. ... and we go to sleep since we processed D3 command and are in the idle tread.

  ---> Nope, device is busy and won't be suspended
6. When you get ack message you clear the device busy state
  
> Or even what if the end-to-end tests don't? Should we risk it because we hesitated to add a safety lock? I think its good to write a code that would be more bug resistant than basing your API on many assumptions. It's not that performance expensive.

I have different opinion. Code should be as simple as possible for the designed task, we always make assumptions about  API behavior, aren't you making assumptions about how state lock works ? And more important IMHO, this lock is not doing what you want here (if I understood it right), the only thing that this lock is doing is preventing that two places call pm_device_action_run() at the same time or that you are preempted in the middle of pm_device_action_run() and you call this function again from the ISR ... and is doing it wrong,  a TOCTOU race can happens.

aborisovich commented:
> > > @aborisovich did you see these comments [#56168 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/56168#discussion_r1218892512)
> > 
> > Yes. Ok, from my perspective, Zephyr System Power Management API exposes `pm_device_action_run()` as part of public API. There are some ideas to share this IPC driver to other architectures not only Intel ADSP. `pm_device_action_run()` function allows to change power state of the driver without using idle thread. That is why those locks are still here. I'm trying to write a more generic code that would be usable also when somebody calls `pm_device_action_run(INTEL_ADSP_IPC_HOST_DEV)`. Then the device would not change power state if it is locked (because it currently performs a transaction).
> > 
> > As for the `device_busy` lock, consider following scenario:
> > 
> > 1. Interrupt from Host with IPC message send to DSP.
> > 2. We schedule IPC answer to Host.
> > 3. Executing IPC send, message is send but receival is not confirmed by the Host.
> > 4. We go back to idle thread waiting for another interrupt from Host that will confirm IPC message receival.
> > 5. ... and we go to sleep since we processed D3 command and are in the idle tread.
> > 
> > This lock prevents this from happening. I understand that "this should not happen since Host responds fast". But what if it don't?
> 
> You are getting it wrong. I am not assuming anything about timing .... I am contesting the efficacy of this lock. You are using a lock to ensure that the device state changing will happen atomically and that is not the case. Following your steps:
> > 1.  Interrupt from Host with IPC message send to DSP
> > 2.  We schedule IPC answer to Host.
> 
>  ---> IPC device is "active" at this point right ? You don't schedule this ipc from inside pm_device_action_run(), correct ?
>        So, what you should do here, is from the function you are scheduling the ipc you set the device busy
> 
> > 3. Executing IPC send, message is send but receival is not confirmed by the Host.
> > 4.  We go back to idle thread waiting for another interrupt from Host that will confirm IPC message receival.
> > 5. ... and we go to sleep since we processed D3 command and are in the idle tread.
> 
>   ---> Nope, device is busy and won't be suspended
> 6. When you get ack message you clear the device busy state
>   
> > Or even what if the end-to-end tests don't? Should we risk it because we hesitated to add a safety lock? I think its good to write a code that would be more bug resistant than basing your API on many assumptions. It's not that performance expensive.
> 
> I have different opinion. Code should be as simple as possible for the designed task, we always make assumptions about  API behavior, aren't you making assumptions about how state lock works ? And more important IMHO, this lock is not doing what you want here (if I understood it right), the only thing that this lock is doing is preventing that two places call pm_device_action_run() at the same time or that you are preempted in the middle of pm_device_action_run() and you call this function again from the ISR ... and is doing it wrong,  a TOCTOU race can happens.
> 
> 
> 
> 

Allright, i got it now. So basically you are saying that PM lock is useless here not the device busy lock. Yeah that makes sense since the busy lock already provides behavior we need. I'll double check this tomorrow and probably remove it. Thanks.

aborisovich commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -142,7 +150,7 @@ bool intel_adsp_ipc_send_message(const struct device *dev,
 	config->regs->idd = ext_data;
 	config->regs->idr = data | INTEL_ADSP_IPC_BUSY;
 	k_spin_unlock(&devdata->lock, key);
-	return true;
+	return 0;
```
No, its not a mistake.
I've fixed all dependencies accepting value from this function.
I'm changing its signature in a4d954bdd01323bc62dfa5ec33910093306f5257 (previously was returning `bool`, `true` value on success. Now returns `int` and error code `0` on success).

marc-hb replied:
> I'm changing its signature in https://github.com/zephyrproject-rtos/zephyr/commit/a4d954bdd01323bc62dfa5ec33910093306f5257 (previously was returning bool, true value on success. Now returns int and error code 0 on success).

I've not been following the whole review sorry but this caught my eye. The return value of `intel_adsp_ipc_send_message()` is being used in SOF, more specifically in `src/drivers/intel/ace/ipc.c`, see SOF commit https://github.com/thesofproject/sof/commit/d5967d19132c#diff-19c1831368f785044dd1175a576d575cfa8c3e5496f66e67ee66e38d9b3a9ad3R152

So this is a backwards-incompatible change that will require careful synchronization in SOF.
BTW it would be good to submit this to `sof/west.yml` CI even before Zephyr merge if you haven't done yet, I mean not just because of this backwards-incompatible change.

PS: I think this comment should stay "unresolved" not because it's unresolved but to keep it visible because it's backwards-incompatible change.

aborisovich commented:
Push update:
* modified 60bc0c104cd68dcaf99308eff8c760bfcdc1b83f `ipc_pm_action()` function so the default return value is `-ENOTSUP`.
This allows to return error code when IPC Device is requested to enter unsupported power states: `PM_DEVICE_ACTION_TURN_OFF` or `PM_DEVICE_ACTION_TURN_ON` .

aborisovich commented:
We are failing tests now. Fw ready does not arrive after resume from D3...

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -145,15 +156,15 @@ bool intel_adsp_ipc_send_message(const struct device *dev,
 	return true;
```
mixing types ... shouldn't it be 0 ?

aborisovich replied:
yes, this was a cause of tests failing thanks

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -125,16 +132,20 @@ bool intel_adsp_ipc_is_complete(const struct device *dev)
+int intel_adsp_ipc_send_message(const struct device *dev,
 			   uint32_t data, uint32_t ext_data)
 {
+	if (pm_device_state_is_locked(INTEL_ADSP_IPC_HOST_DEV))
```
missing braces

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -125,16 +132,20 @@ bool intel_adsp_ipc_is_complete(const struct device *dev)
 			   uint32_t data, uint32_t ext_data)
 {
+	if (pm_device_state_is_locked(INTEL_ADSP_IPC_HOST_DEV))
+		return -ECANCELED;
```
`-EAGAIN`

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -145,15 +156,15 @@ bool intel_adsp_ipc_send_message(const struct device *dev,
+	int ret = intel_adsp_ipc_send_message(dev, data, ext_data);
 
-	if (ret) {
+	if (ret != 0 && ret != -ECANCELED) {
```
you have changed the behavior here, if `intel_adsp_ipc_send_message()` returns `-EBUSY` you will execute this block ... 
The right is probably `if (ret == 0)`

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -207,12 +218,123 @@ static int dt_init(const struct device *dev)
+		(const struct ipc_control_driver_api *)dev->api;
+
+	/* Return as default value when given PM action is not supported */
+	int ret = -ENOTSUP;
```
This is probably causing problems  .... if `PM_DEVICE_ACTION_SUSPEND` and `api->suspend_fn == NULL` this function will return `ENOTSUP` which is wrong.

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -125,16 +132,20 @@ bool intel_adsp_ipc_is_complete(const struct device *dev)
+int intel_adsp_ipc_send_message(const struct device *dev,
 			   uint32_t data, uint32_t ext_data)
 {
+	if (pm_device_state_is_locked(INTEL_ADSP_IPC_HOST_DEV))
```
Thinking about it ... I don't think that is what you really want to check here ... What happens if the device in not locked but it is suspended ?? I think what you want to check here is that the device is active, so you should use `pm_device_state_get()`.

aborisovich replied:
Added check for device active state.

ceolin commented on soc/xtensa/intel_adsp/common/ipc.c:
```diff
@@ -207,12 +218,123 @@ static int dt_init(const struct device *dev)
+{
+	if (k_is_in_isr()) {
+		return -ESRCH;
+	}
```
Do you really need this ??

aborisovich replied:
Well, you started a subject on TOCTOU race and did not answer my question... Please read my comments above.

ceolin replied:
Nope, it does not solve it. If you have two threads that call `pm_device_action_run()` is possible that both threads pass the check `pm_device_state_is_locked()` and both ends up calling `ipc_pm_action` and in that case you don't know exactly what will happen. That is a problem that device_runtime solves, pm_device_state_lock() was not thought to be used this way (though there is nothing disallowing it). It was simply not thought.

My suggestion is to remove it for now (since is not really protecting) and ensure that you are calling it from a safe context. I will probably change `pm_device_lock()` to return a boolean and fails when the device state is already locked and you can use here it and that will probably be a solution.

Does it work for you ?

ceolin commented on subsys/logging/backends/log_backend_adsp_hda.c:
```diff
@@ -342,20 +344,20 @@ void adsp_hda_log_cavstool_hook(uint32_t written)
+		if (ret == -ECANCELED) {
+			return;
+		}
+	} while (ret);
```
There is a bug here ... you changed `intel_adsp_ipc_send_message` signature but in success it is still returning `true`, so you will loop forever in case the message is sent.

ceolin reviewed (changes requested):
Look inline comments please.

ceolin commented on subsys/logging/backends/log_backend_adsp_hda.c:
```diff
@@ -342,20 +344,20 @@ void adsp_hda_log_cavstool_hook(uint32_t written)
 					     (written << 8) | CHANNEL);
-	} while (!done);
-
+		if (ret == -ECANCELED) {
```
shouldn't it be `-ESHUTDOWN` ?

ceolin commented on soc/xtensa/intel_adsp/common/include/intel_adsp_ipc.h:
```diff
@@ -135,30 +139,32 @@ bool intel_adsp_ipc_is_complete(const struct device *dev);
- * in progress (in the sense of intel_adsp_ipc_is_complete()).
+ * Returns 0 if the message was sent, negative error values:
+ * -EBUSY if there is already IPC message processed (intel_adsp_ipc_is_complete returns false).
+ * -ECANCELED if IPC device will not send the message as it undergoes power
```
`-ECANCELED` or `-ESHUTDOWN` ?

ceolin commented:
@aborisovich  please fix the return error code information that I commented and if you agree just drop that check if it is from ISR (though I won't block this commit if you decided to keep that).

Once you fix that error code mismatch it is good to me.

aborisovich commented:
Push update - addressed @ceolin issues found:
* fixed return codes (-ESHUTDOWN is used where it should).
* removed check `k_is_in_isr()` from the `ipc_pm_action` along with the documentation that the action is not supported in ISR.

tmleman commented:
@aborisovich SOF CI looks good https://github.com/thesofproject/sof/pull/7325. How did our internal full scope tests go?

aborisovich commented:
> @aborisovich SOF CI looks good [thesofproject/sof#7325](https://github.com/thesofproject/sof/pull/7325). How did our internal full scope tests go?

We have improvement of ~1.5% passrate.
New tests passing:
`hda_codec_loopback_rtd3` ✔️
`channel_remapping_mode_dynamic` ✔️ 
`mixers_max_inputs` ✔️ 
`cpa_toggle_without_fw_logs` ✔️ 
Other D3 related tests are apssing too.
There is regression in 1 test: `load_fw_stress_graceful_shutdown` ❌ 
I'll investigate on which iteration we have failures.
I'll restart full scope tests today after rebase to newest zephyr main.

aborisovich commented:
Last run after Zephyr main rebase gave us ~3% improvement.
Comparing to previous run, new tests started to pass:
`ace_generic_dmic` ✔️ 
`ace_raw_capture` ✔️ 
`kd_topology_dmic` ✔️ 
Unfortunately, some of the:
`hda_codec_loopback_rtd3` tests failed with glitches after D3.
What keeps me wonder is why only the half of the tests in this group failed and second half passed...
